### ISSUE TITLE
Creating views after re-opening DB is *way* slower than doing it before first close.

### ISSUE THREAD

**Issue by stefanwascoding** [couchbase/couchbase-lite-ios#2079]
Adding a view after a database has been created has a serious performance penalty compared to adding it on creation.

This works as expected:

- create DB
- add 50.000 random docs
- create view
- close DB
- open DB
- use view & limit to 1 doc - response time is ~0.5ms

in contrast:

- create DB
- add 50.000 random docs
- close DB
- open DB
- create view
- use view & limit to 1 doc - response time is >**600ms**

In fact, when using no limit, it only takes about ~100ms additional time to fetch the whole view, and not just one entry.
( I did run the view multiple times, so it was up-to-date, before taking the time. )

From what I can tell from the resulting *db.sqlite* files, the "fast" ones, contain a `maps_1` table and a corresponding entry in the `sqlite_stat1` table for that, but the "slow" ones only contain the `maps_1` table.

Example project: [Slow View.zip](https://github.com/couchbase/couchbase-lite-ios/files/1768968/Slow.View.zip)

--------
* Version:   1.4.1
* Client OS: iOS 11.2 (Simulator & Device)

**Comment by snej** [comment]
I'm guessing that in the fast case most of the database is already in SQLite's RAM cache when the view index is being generated.
Try running your test in Instruments with the CPU profiler, to see where the time is going.

**Comment by stefanwascoding** [comment]
I've modified the example project to split this in two, just to make sure ... on the first run it creates both databases, and on the second run it runs the queries: [Slow_View_2.zip](https://github.com/couchbase/couchbase-lite-ios/files/1770988/Slow_View_2.zip) (if you want to test it yourself, you will need to change the path where it creates the databases - ViewController.m:37).

The instruments trace (for both runs): [CBL-slow-query.trace.zip](https://github.com/couchbase/couchbase-lite-ios/files/1770989/CBL-slow-query.trace.zip)

And some screenshots:
<img width="1597" alt="testfast" src="https://user-images.githubusercontent.com/2292035/36843188-6e928f9c-1d4e-11e8-9eb4-24ac3087b664.png">
<img width="1597" alt="testslow" src="https://user-images.githubusercontent.com/2292035/36843189-6eb56526-1d4e-11e8-8a75-5c4ece2e6883.png">

The fast path seems to quick for Instruments to properly pick up on unfortunately ... and the slow path seems to spend most of its time in SQLite code, and I have no experience with that. Maybe you can see what's going on.

**Comment by stefanwascoding** [comment]
Tried to get more traces from the fast path:

<img width="1421" alt="fast_1" src="https://user-images.githubusercontent.com/2292035/36843567-a2ebb5a6-1d4f-11e8-8322-e5ba0f373fb1.png">
<img width="1421" alt="fast_2" src="https://user-images.githubusercontent.com/2292035/36843568-a307ba12-1d4f-11e8-9478-445a004a1a90.png">

**Comment by snej** [comment]
Your fast test queries the view immediately after populating the database (contrary to what you described above); this causes the view to be indexed. Since the docs have just been added, SQLite's buffer caches are hot, so there's very little I/O involved.
In the slow test, the view isn't queried until after the db is reopened, meaning that SQLite has nothing in cache yet and has to read the entire DB from disk.
In other words, this behavior is about what I'd expect.

**Comment by stefanwascoding** [comment]
But the times I quoted don't come from the first run of the app, when the db gets populated.

So basically what the example app is supposed to to:
- First Run - db creation

  1. create a "fast" database, by opening a db, filling it with documents, query view "x", close db (don't care about these times)
  2. create a "slow" database, by opening a db, filling it with documents, close db  (no view yet)

- Second Run - db use, and timing test

  1. open "fast" database, query view "x" _multiple times_, average response time for 1 document <0.5ms, close db
  2. open "slow" database, query view "x" _multiple times_, average response time for 1 document >600ms, close db

- Third Run & later
  Same as 2nd run

Caches should not play a role here when the behaviour persists across restart of the app, should they? Also, in the slow test the performance never gets better, no matter if you do the query multiple times in one run, or when doing multiple runs.

I can even take the resulting cblite2 files to another computer, and when I re-time the views there, they show the same performance characteristics.

**Comment by stefanwascoding** [comment]
Side note: When I add a new view to either the fast or slow database, after a few runs, they will *both* be slow (and don't improve over multiple queries (after the initial one) or app runs).

**Comment by stefanwascoding** [comment]
Even more minimal code for macOS this time: 
[View_Test_macOS.zip](https://github.com/couchbase/couchbase-lite-ios/files/1772719/View_Test_macOS.zip)

If you run the app the first time, it will create two files on the Desktop:
- `fast.cblite2` by doing `create - seed - ............... - query view - close`
- `slow.cblite2` by doing `create - seed - close - re-open - query view - close`

When you run the app a second (or third, ...) time, it will open each of those files in turn, and do two queries of the view that was created in the first run, the first query just to warm up any caches, and the second query for measurement ... and the results are consistent in their relative difference between runs (and on different computers):

```
[...] CBL view speed test[41397:736806] query in fast db took   0.15ms (1 results)
[...] CBL view speed test[41397:736806] query in slow db took 109.79ms (1 results)
```

**Comment by snej** [comment]
OK, that is weird. From the logs, the difference is that the SQL query for the view query is a lot slower in the "slow" db. I don't know why that would be, or why it persists. 

You mentioned something about the `sqlite_stat1` table -- that's internal to SQLite. It stores statistics about the rows in a table, and is used for query optimization. IIRC it's created by the `PRAGMA ANALYZE` command, and also IIRC, CBL 1.x runs that pragma at some point after the db is created. (It might be on the first close?) So the difference might be due to the fact that SQLite was able to analyze the view contents. And the fix might be to run the ANALYZE command again at some point, although I don't know what a good heuristic for that would be.

We're probably not going to analyze or fix this ourselves, since CBL 1.x is in maintenance mode and this isn't a critical issue. CBL 2 doesn't use views at all, so this issue wouldn't apply there. If you come up with a reasonable fix we'd consider taking it, although in my experience this sort of thing can be tricky because it might end up making your test faster at the expense of making other user cases slower...

**Comment by stefanwascoding** [comment]
Thanks for the tip!

The problem goes away when I manually force it to analyze the view after it has been computed, and before doing the benchmark tests.

Side note: `optimizeSQLIndexes` is called on DB-open (and even then only when the DB changed significantly), although https://www.sqlite.org/lang_analyze.html suggests it should be done on DB-close, when the DB connection has collected some usage data.

**Comment by stefanwascoding** [comment]
I'm building against the CBL source, so it wasn't hard to modify it to trigger the optimization when I want it to run.

**Comment by snej** [comment]
With a new-enough SQLite it's preferable to run PRAGMA OPTIMIZE. We do this in CBL 2.

>To achieve the best long-term query performance without the need to do a detailed engineering analysis of the application schema and SQL, it is recommended that applications run "PRAGMA optimize" (with no arguments) just before closing each database connection. Long-running applications might also benefit from setting a timer to run "PRAGMA optimize" every few hours.
